src/main/RemoteRequest.js

```diff
@@ -0,0 +1,87 @@
+/**
+ * RemoteRequest is used to for http requests on a remote server which can be
+ * fetched in chunks, e.g. using a Range request.
+ * @flow
+ */
+'use strict';
+
+import Q from 'q';
+import ContigInterval from './ContigInterval';
+
+class RemoteRequest {
+  url: string;
+  cache: Object;
+  numNetworkRequests: number;  // track this for debugging/testing
+
+  constructor(url: string) {
+    this.cache = require('memory-cache');
+    this.url = url;
+    this.numNetworkRequests = 0;
+  }
+
+  get(contig: string, start: number, stop: number): Q.Promise<Object> {
+    var length = stop - start;
+    if (length <= 0) {
+      return Q.reject(`Requested <0 interval (${length}) from ${this.url}`);
+    }
+
+    // First check the cache.
+    var contigInterval = new ContigInterval(contig, start, stop);
+    var buf = this.cache.get(contigInterval);
+    if (buf) {
+      return Q.when(buf);
+    }
+
+    // Need to fetch from the network.
+    return this.getFromNetwork(contig, start, stop);
+  }
+
+  /**
+   * Request must be of form "url/contig?start=start&end=stop"
+  */
+  getFromNetwork(contig: string, start: number, stop: number): Q.Promise<Object> {
+    var length = stop - start;
+    if (length > 50000000) {
+      throw `Monster request: Won't fetch ${length} bytes from ${this.url}`;
+    }
+    var xhr = new XMLHttpRequest();
+    var endpoint = this.getEndpointFromContig(contig, start, stop);
+    xhr.open('GET', endpoint);
+    xhr.responseType = 'json';
+    xhr.setRequestHeader('Content-Type', 'application/json');
+
+    return this.promiseXHR(xhr).then(json => {
+      // extract response from promise
+      var buffer = json[0];
+      var contigInterval = new ContigInterval(contig, start, stop);
+      this.cache.put(contigInterval, buffer);
+      return buffer;
+    });
+  }
+
+  getEndpointFromContig(contig: string, start: number, stop: number): string {
+    return this.url + "/" + contig + "?start=" + start + "&end=" + stop;
```

Comment:
how about making this a bit more consice with [ES6's template feature](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals)? e.g.:

```javascript
   return `${this.url}/${contig}?start=${start}&end=${stop}`;
```